sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/DocumentQueryExecutionContextFactory.java

```diff
@@ -370,75 +419,90 @@ public static <T> Flux<? extends IDocumentQueryExecutionContext<T>> createSpecia
+                if (maxLimit > maxItemSizeForVectorSearch) {
+                    throw new NonStreamingOrderByBadRequestException(HttpConstants.StatusCodes.BADREQUEST,
+                        "Executing a vector search query with TOP or LIMIT larger than the maxItemSizeForVectorSearch " +
+                            "is not allowed");
+                }
+                // Set initialPageSize based on the smallest of TOP or LIMIT
+                if (queryInfo.hasTop() || queryInfo.hasLimit()) {
+                    int pageSizeWithTopOrLimit = Math.min(queryInfo.hasTop() ? queryInfo.getTop() : Integer.MAX_VALUE,
+                        queryInfo.hasLimit() && queryInfo.hasOffset() ?
+                            queryInfo.getLimit() + queryInfo.getOffset() : Integer.MAX_VALUE);
+                    initialPageSize = pageSizeWithTopOrLimit;
                 }
             }
-            // TODO: do not support continuation in string format right now
-            //            else if (isContinuationExpected)
-            //            {
-            //                if (initialPageSize < 0)
-            //                {
-            //                    initialPageSize = (int)Math.Max(feedOptions.MaxBufferedItemCount,
-            //                      ParallelQueryConfig.GetConfig().DefaultMaximumBufferSize);
-            //                }
-            //
-            //                initialPageSize = Math.Min(
-            //                    (int)Math.Ceiling(initialPageSize / (double)targetRanges.Count) * PageSizeFactorForTop,
-            //                    initialPageSize);
-            //            }
-        }
 
-        List<FeedRangeEpkImpl> feedRangeEpks = targetRanges.stream().map(FeedRangeEpkImpl::new)
-                                                   .collect(Collectors.toList());
+            // We need to compute the optimal initial page size for order-by queries
+            if (queryInfo.hasOrderBy()) {
+                int top;
+                if (queryInfo.hasTop() && (top = queryInfo.getTop()) > 0) {
+                    int pageSizeWithTop = Math.min(
+                        (int) Math.ceil(top / (double) targetRanges.size()) * PageSizeFactorForTop,
+                        top);
 
-        if (collection.getPartitionKey() != null && cosmosQueryRequestOptions.getPartitionKey() != null
-            && collection.getPartitionKey().getKind().equals(PartitionKind.MULTI_HASH)) {
-            feedRangeEpks = resolveFeedRangeBasedOnPrefixContainer(feedRangeEpks, collection.getPartitionKey(),
-                cosmosQueryRequestOptions.getPartitionKey());
+                    if (initialPageSize > 0) {
+                        initialPageSize = Math.min(pageSizeWithTop, initialPageSize);
+                    } else {
+                        initialPageSize = pageSizeWithTop;
+                    }
+                }
+                // TODO: do not support continuation in string format right now
```

Comment:
WHat about this comment?

Reply:
I'm not sure about this comment. This is an older comment, which already there in the code. Not related to my PR.